share/man/man8/man8.i386/boot.8
.\" Copyright (c) 1993 Berkeley Software Design, Inc. All rights reserved.
.\" The Berkeley Software Design Inc. software License Agreement specifies
.\" the terms and conditions for redistribution.
.\"	BSDI $Id: boot.8,v 1.2 1994/02/03 02:02:14 polk Exp $
.\"
.Dd March 2, 1993
.Dt BOOT 8 i386
.Os
.Sh NAME
.Nm boot
.Nd bootstrap procedures for
.Tn BSD Ns /386
.Sh DESCRIPTION
.Tn BSD Ns /386
is normally booted by simply turning on the power to the
computer and waiting.  Under normal conditions, the floppy drive will
be empty, so the BIOS will read and execute the boot block from the
hard disk, which will read and execute the program
.Pa /boot .
Then,
.Pa /boot
will normally read and execute the kernel,
.Pa /bsd .
Next, the kernel does a series of probes to determine what hardware is
present, then executes the program
.Xr init 8 .
Init then runs the shell script
.Pa /etc/rc .
This script checks the file systems, configures the network hardware
and starts background system tasks.  If it completes normally, init
will spawn getty processes for the terminals listed in
.Pa /etc/ttys ,
to allow users to log in.
If 
.Pa /etc/rc
exits with an error code, init will instead run a single user shell.
.Pp
A number of these steps may be configured or modified
during a single execution, or in configuration files.
.Pp
The program
.Pa /boot
prints the kernel text, data and bss sizes while loading
the kernel, then prints the entry location, pauses briefly,
and then transfers control to the kernel.
This process takes several seconds.
During this time, if you type a key on the keyboard, the load
will be aborted, and you will be prompted with:
.Pp
.Dl Boot:
.Pp
At this point, you can specify an alternate kernel to boot,
potentially on a different device.
The default boot flags for the kernel are changed
to request a boot to single user, possibly asking for a root filesystem
location (see below).
.Pp
The complete boot command is:
.Pp
.Bd -ragged -offset indent
.Ar dev\| Ns ( Ar adapter , Ar controller ,
.Ar unit , Ar partition ) Ns Ar pathname
.Op Fl adrsw | Ar howto
.Ed
.Pp
or
.Pp
.Bd -ragged -offset indent
.Fl \& Ns Ar bootcommand
.Ed
.Pp
If the command begins with
.Fl \&
the line is treated as a boot command (see below),
otherwise the line is used as a boot specification.
The device specification
.Pf ( Ar dev
through the closing parenthesis) is optional;
if omitted, the kernel is loaded from the device and partition from which
the bootstrap was loaded.
Otherwise,
.Ar dev
is the type of the desired device:
.Bd -ragged -offset indent
.Bl -column "eahaxx"
.It Li fd No "	floppy disk"
.It Li wd No "	traditional hard disks"
.It Li sd No "	SCSI hard disks"
.It Li aha No "	disk on Adaptec ISA SCSI host adapter"
.It Li eaha No "	disk on Adaptec EISA SCSI host adapter"
.El
.Ed
.Pp
The latter two names are useful when more than one type of host adapter
is present.
.Pp
The values
inside the parentheses should be numbers.
.Ar adapter
is normally unused.
The disk controller and disk unit numbers
are given by
.Ar controller
and
.Ar unit .
The
.Ar partition
value refers to the disk
partitions by number (0=a, 1=b, etc).
If fewer than four numbers are specified, then zeroes are prepended.
For example, a single number sets the partition,
leaving the other parameters at 0.
If nothing appears between the parentheses,
then all four parameters are set to 0.
.Pp
The pathname identifies the desired file to be booted.
If no device specification or pathname is present
(that is, only a RETURN is typed),
the default boot sequence is initiated.
The default boot command is
.Dq Li /bsd .
.Pp
The following options to the loaded kernel are available.
With no options, the default is the same as specifying the 
.Fl as
options if the boot was interrupted, and no options if the bootstrap was
automatic.
If any options are present, only those present are actually used; a
.Fl \&
alone explicitly disables all options.
.Bl -tag -width indent
.It Fl a
Askname; if loading a generic kernel, the kernel will prompt
for the name of the root filesystem device.
If loading a bootstrap program, prompt for a boot command.
.It Fl d
Enable debugging.
Currently, this enables diagnostic output during the startup operations
of the kernel.
.It Fl r
Use the default (compiled-in) root filesystem
rather than using the load device as the root.
.It Fl s
Boot to a single-user shell rather than checking filesystems
and going into multiuser operation.
.It Fl w
Mount the root filesystem read/write rather than read-only (not yet
implemented in the kernel).
.El
.Pp
The boot flags may also be set using the
.Ar howto
parameter, which is a decimal, octal or hex number that sets the
boot flags \(em see the RB_* constants in
.Pa sys/reboot.h .
The use of the flags above is generally recommended rather than a numeric value.
.Pp
If a bootstrap command line begins with
.Fl \& ,
it is treated as an internal boot command.
There is one such command currently:
.Fl autodebug Ar number
sets the level of debugging for autoconfiguration to the specified
.Ar number .
The levels available are 0 (no debugging), 1 (print information about each
device and location probed and the result, pausing after each screenful),
and 2 (print information as for 1, but confirm whether each device should
be probed).
Level 2 is useful if probing some device location causes a failure.
.Pp
If the file
.Pa /etc/boot.default
exists, the
.Pa /boot
program reads a single line from that file for the name and location
of the kernel to be booted, as well as any options.
The file contents take the same form as the response to the
.Li Boot:
prompt.
This line becomes the default kernel and options;
if loading of the kernel is interrupted,
and then just RETURN is typed, this default is executed.
.Sh BOOT HINTS
Here are a few common boot procedures.  The sections that follow
describe what is happening in more detail.
.Pp
To boot a single user shell from the hard disk: make sure the floppy
drive is empty, reset the processor, type space while the kernel is
being loaded, then type a RETURN at the Boot: prompt.
You should
usually run fsck to check the file systems before doing anything else.
If you type control-D, the system will come up multi-user.  Don't
forget that the root is read-only at the start, so if you want to do
maintenance in single-user mode, you need to do
.Pp
.Dl mount -u /
.Pp
to enable writes.
.Pp
To boot a backup kernel from the hard disk: make sure the floppy
drive is empty, reset the processor, type space while the kernel is
being loaded, then type the desired kernel name at the Boot: prompt
(e.g. bsd.good).  This will land you in a single user shell.
.Pp
To boot
.Tn BSD Ns /386
from the hard disk, even though the FDISK label is set
to boot DOS by default: insert a
.Tn BSD Ns /386
boot floppy and reset the
processor.  At the Boot: prompt, type
.Dq Li wd()bsd -
(replace
.Li wd
with
.Li sd
for a scsi main disk).  Leave off the
.Li \&-
if you want a single user shell.
The floppy drive is now free, and can be used normally when the system
finishes booting.  (See below for creating a floppy that boots
straight
.Tn BSD Ns /386
without any operator intervention.)
.Pp
To boot with a CD-ROM root: insert a
.Tn BSD Ns /386
boot floppy and a
.Tn BSD Ns /386
CD-ROM and reset the processor.  At the Boot: prompt, press RETURN.
Type RETURN again at when offered the chance to change floppies.
After kernel autoconfiguration, you will be prompted
.Dq Li "root device?" .
Type
.Dq Li sd0a
for a scsi CD-ROM, or
.Dq Li mcd0a
for the Mitsumi ISA
CD-ROM.  The boot floppy is no longer needed, so you may insert
another floppy, if desired.
.Pp
If there is a floppy in drive 0 when the processor is reset, then its
boot block with be read and executed by the BIOS.  The floppy boot
always starts at the Boot: prompt.  If you just type a RETURN, the
default action is to boot /bsd from the floppy.  Since standalone
floppy systems are cramped for space, the floppy boot gives you a
chance to change media between loading the kernel and starting it.
Therefore, the disk you actually use as the root file system need not
have a kernel binary.
.Pp
If you boot with a floppy root, then you must not remove the floppy
while running; however, if you specify an alternate device, then the
floppy is free and can be used for any purpose.
.Pp
To make a floppy that boots straight to
.Tn BSD Ns /386
on a hard disk, make
a copy of your distribution boot floppy, and in the copy, put a line like
.Pp
.Dl wd(0,0)/bsd
.Pp
in the file
.Pa /etc/boot.default .
Replace
.Li wd
with
.Li sd
if you have a scsi main disk;
change the unit numbers as needed.
.Sh BOOT DETAILS
The boot process on a PC consists of a number of stages.
.Pp
The first stage is handled by the system BIOS.
It selects a boot device, then loads and executes the first sector
of the device.
The device selection is sometimes configurable, but often defaults
to the
.Dq A
floppy if a diskette is present, otherwise the
.Dq C
(primary) hard disk.
If the disk is a hard disk, the first block may be a
.Tn BSD Ns /386
boot block, a standard PC master boot record, or a program
like
.Xr bootany
(see
.Xr disksetup 8 ) ,
which allows a choice of systems to be booted.
In the latter two cases, a PC FDISK partition table in the first block
lists the PC-style partitions and indicates which should normally be booted.
Another bootstrap facility may also be used, such as the OS/2
boot manager.
If a
.Tn BSD Ns /386
partition is selected,
a
.Tn BSD Ns /386
boot block
is then executed
(usually one of
.Pa fdboot ,
.Pa wdboot ,
.Pa ahaboot ,
or
.Pa eahaboot ) .
The boot block loads a slightly larger program from the following
15 sectors
(usually one of
.Pa bootfd ,
.Pa bootwd ,
.Pa bootaha ,
or
.Pa booteaha ) .
That program prints
.Dq Li "Loading /boot"
and then loads and executes
.Pa /boot
from the indicated partition.
.Pa /boot
then loads and executes the kernel, passing the boot flags and additional
information to the kernel.
.Sh FILES
.Bl -tag -width /etc/boot.default
.It Pa /boot
Second stage boot
.It Pa /sbin/init
The first user program
.It Pa /etc/rc
Startup shell script
.It Pa /etc/boot.default
Default boot command
.Sh SEE ALSO
.Xr disklabel 8 ,
.Xr disksetup 8 ,
.Xr init 8 ,
.Xr reboot 8 ,
.Xr shutdown 8


sys/i386/stand/boot.c
/*-
 * Copyright (c) 1992, 1993 Berkeley Software Design, Inc. All rights reserved.
 * The Berkeley Software Design Inc. software License Agreement specifies
 * the terms and conditions for redistribution.
 *
 *	BSDI $Id: boot.c,v 1.16 1994/01/05 18:46:24 karels Exp $
 */

/*-
 * Copyright (c) 1990 The Regents of the University of California.
 * All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * William Jolitz.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1990 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)boot.c	7.3 (Berkeley) 5/4/91";
#endif /* not lint */

#include "param.h"
#include "reboot.h"
#include <a.out.h>
#include <setjmp.h>
#include <string.h>
#include <i386/isa/isa.h>
#include "saio.h"

/*
 * Boot program... arguments from lower-level bootstrap determine
 * whether boot stops to ask for system name and which device
 * boot comes from.
 */

#define MAXBOOT		(BRELOC - IOM_SIZE - 8*1024)
#define FLOPPYDEV 2	/* from conf.c */

char	line[100];
char	default_boot[] = DEFAULTBOOT;
extern	int opendev, bootdev, cyloffset;
extern	jmp_buf exception;
extern	int bootdebug;

struct bootparams {
	struct	bootparamhdr hdr;
	u_char	data[BOOT_MAXPARAMS];
} bootparams = {
	{ BOOT_MAGIC, sizeof(struct bootparamhdr) },
};
caddr_t ebootparams = (caddr_t) &bootparams + sizeof(struct bootparamhdr);

void boot_file __P((int io, int howto));
void makebootline __P((void));

void
main(howto, dev, off, bhdr)
	int howto, dev, off;
	struct bootparamhdr bhdr;
{
	int io;
	int n;
	char *cp;
	char *defboot = default_boot, defbuf[512];
	char *fp = NULL, *ep, *lp;
	struct bootparam *bp;

	if ((dev&B_MAGICMASK) == B_DEVMAGIC) {
		bootdev = dev;
		cyloffset = off;
		makebootline();
#ifdef CHANGEFLOPPY
		if (B_TYPE(bootdev) == FLOPPYDEV)
			howto = RB_SINGLE|RB_ASKNAME;
#endif
	} else {
		howto = RB_SINGLE|RB_ASKNAME;
		cyloffset = 0;
	}
	if (bhdr.b_magic == BOOT_MAGIC) {
		setbootparams(&bhdr);
		for (bp = B_FIRSTPARAM(&bootparams.hdr); bp;
		    bp = B_NEXTPARAM(&bootparams.hdr, bp))
		    	switch (bp->b_type) {
			case B_BIOSGEOM:
				setbiosgeom((struct biosgeom *) B_DATA(bp));
				break;
#ifdef BSDGEOM
			case B_BSDGEOM:
				setbsdgeom((struct bsdgeom *) B_DATA(bp));
				break;
#endif
			default:
				break;
		}
	}

	/*
	 * If the file /etc/boot.default exists,
	 * we read it for the default boot path.
	 * We do minimal checking; if the file is not empty,
	 * we do what it says.
	 */
	if ((io = open("/etc/boot.default", 0)) >= 0) {
		if ((n = read(io, defbuf, sizeof(defbuf))) > 0) {
			fp = defbuf;
			ep = fp + n;
		}
		close(io);
	}
	if (_setjmp(exception)) {
		close(io);
		printf("- load aborted\n");
		howto = RB_SINGLE|RB_ASKNAME;
		cyloffset = 0; 
		fp = NULL;
	}
		
	for (;;) {
		line[0] = '\0';
		lp = line;
		if (fp) {
			lp = fp;
			if (*fp != '-') {
				defboot = fp;
				howto = RB_AUTOBOOT;
			}
			for (; fp < ep; fp++) {
				if (*fp == '\n') {
					*fp++ = 0;
					break;
				}
			}
			if (fp >= ep)
				fp = NULL;
			printf("Boot: %s\n", lp);
		} else if (howto & RB_ASKNAME) {
			printf("Boot: ");
			gets(line);

			cyloffset = 0;
		}

		if (*lp == 0) {
			lp = defboot;
			printf("Boot: %s\n", lp);
		}

		if (*lp == '-') {
			dobootopts(lp);
			continue;
		}

		/* process additional flags if any */
		if (cp = (char *)index(lp, ' ')) {
			if (bootflags(cp, &howto) == -1) {
				howto |= RB_ASKNAME;
				continue;
			}
			*cp = '\0';
		}

		if ((io = open(lp, 0)) < 0)
			printf("can't open %s: %s\n", lp,
			    strerror(errno));
		else
			boot_file(io, howto);

		howto = RB_SINGLE|RB_ASKNAME;
		cyloffset = 0; 
	}
}

/*
 * Process boot flags.
 * Supports either getopt-style boolean flags, or a single numeric parameter,
 * with either specifying the complete flags if present (ignoring previous).
 * Use original flags if no flags are actually present.
 * Could use getopt, but would have to construct argc/argv and do most of
 * these checks anyway.
 */
bootflags(cp, pflags)
	char *cp;
	int *pflags;
{
	int first = 1;
	int flags = 0;

	while (*cp) {
		while (*cp == ' ')
			cp++;
		if (*cp == '\0')
			break;
		if (*cp == '-') {
			first = 0;
			while (*++cp)
				switch (*cp) {
				case ' ':
					goto nextarg;
				case 'a':
					flags |= RB_ASKNAME;
					break;
				case 'd':
					flags |= RB_KDB;
					break;
				case 'r':
					flags |= RB_DFLTROOT;
					break;
				case 's':
					flags |= RB_SINGLE;
					break;
				case 'w':
					flags |= RB_RWROOT;
					break;
				default:
					goto usage;
			}
			continue;
		}
		if (first && *cp >= '0' && *cp <= '9') {
			*pflags = strtol(cp, 0, 0);
			return (0);
		}
		goto usage;

	nextarg: ;
	}
	if (first == 0)
		*pflags = flags;
	return (0);

usage:
	printf("usage: boot-file [ -adrsw ]\n");
	return (-1);
}

dobootopts(cp)
	char *cp;
{
	struct bootparam param;
	int val;

	if (strncmp(cp, "-autodebug", sizeof("-autodebug" - 1)) == 0) {
		if (cp = (char *)index(cp, ' ')) {
			val = strtol(cp, 0, 0);
			param.b_type = B_AUTODEBUG;
			param.b_len = sizeof(param) + sizeof(val);
			addbootparam(&param, (void *) &val);
		} else
			printf("usage: -autodebug number\n");
		return;
	}
	if (strncmp(cp, "-bootdebug", sizeof("-bootdebug" - 1)) == 0) {
		if (cp = (char *)index(cp, ' '))
			bootdebug = strtol(cp, 0, 0);
		else
			printf("usage: -bootdebug number\n");
		return;
	}
}

setbootparams(bhp)
	struct bootparamhdr *bhp;
{

	if (bhp->b_len <= sizeof(bootparams)) {
		bcopy(bhp, &bootparams, bhp->b_len);
		ebootparams = (caddr_t) &bootparams + bootparams.hdr.b_len;
	} else
		printf("setbootparams: parameters too large\n");
}

addbootparam(bp, val)
	struct bootparam *bp;
	void *val;
{

	if (ebootparams + bp->b_len <=
	    (caddr_t) &bootparams + sizeof(bootparams)) {
		bcopy(bp, ebootparams, sizeof(*bp));
		bcopy(val, ebootparams + sizeof(*bp), bp->b_len - sizeof(*bp));
		ebootparams += bp->b_len;
		bootparams.hdr.b_len += bp->b_len;
	} else
		printf("boot parameters too large\n");
}

void
boot_file(io, howto)
	int io, howto;
{
	struct exec x;
	int i;
	char *addr, c;

	i = read(io, (char *)&x, sizeof x);
	if (i != sizeof x ||
	    (x.a_magic != 0407 && x.a_magic != 0413 && x.a_magic != 0410)) {
		printf("Bad format\n");
		return;
	}

	if (x.a_text + x.a_data >= MAXBOOT) {
		printf("File too large (%dK limit)\n", MAXBOOT / 1024);
		return;
	}
		
	printf("%d", x.a_text);
	if (x.a_magic == 0413 && lseek(io, CLBYTES, 0) == -1)
		goto shread;
	if (x.a_text > IOM_BEGIN) {
		if (read(io, (char *)0, IOM_BEGIN) != IOM_BEGIN)
			goto shread;
		if (read(io, (char *)IOM_END, x.a_text - IOM_BEGIN) !=
		    x.a_text - IOM_BEGIN)
			goto shread;
	} else if (read(io, (char *)0, x.a_text) != x.a_text)
		goto shread;

	addr = (char *)x.a_text;
	if (addr >= (char *)IOM_BEGIN)
		addr += IOM_SIZE;
	if (x.a_magic == 0413 || x.a_magic == 0410)
		while ((int)addr & CLOFSET)
			*addr++ = 0;
	if (addr == (char *)IOM_BEGIN)
		addr == (char *)IOM_END;
	printf("+%d", x.a_data);
	if (addr < (char *)IOM_BEGIN && addr + x.a_data > (char *)IOM_BEGIN) {
		if (read(io, addr, IOM_BEGIN - (int)addr) !=
		    IOM_BEGIN - (int)addr)
			goto shread;
		if (read(io, (char *) IOM_END,
		    x.a_data - (IOM_BEGIN - (int)addr)) !=
		    x.a_data - (IOM_BEGIN - (int)addr)) 
			goto shread;
	} else if (read(io, addr, x.a_data) != x.a_data)
		goto shread;

	addr += x.a_data;
	if (addr >= (char *)IOM_BEGIN)
		addr += IOM_SIZE;
	printf("+%d", x.a_bss);
	/*
	 * Kludge: if program including bss is too large, don't clear bss;
	 * currently only kernels are that large, and the kernel clears bss.
	 * (So does srt0, at least currently)
	 */
	if (x.a_text + x.a_data + x.a_bss >= MAXBOOT)
		x.a_bss = MAXBOOT - (x.a_text + x.a_data);
	if (addr < (char *)IOM_BEGIN && addr + x.a_bss > (char *)IOM_BEGIN) {
		bzero(addr, IOM_BEGIN - (int)addr);
		bzero((char *)IOM_END, x.a_bss - (IOM_BEGIN - (int)addr));
	} else
		bzero(addr, x.a_bss);

	/* mask high order bits corresponding to relocated system base */
	x.a_entry &= 0x00ffffff;
	printf(" start 0x%x", x.a_entry);

	/* pause briefly in case user wants to abort */
	for (i = 0; i < 100000; i++)
		if (c = scankbd())
			_longjmp(exception,1);
	printf("\n");

	close(io);		/* after last longjmp */
#ifdef CHANGEFLOPPY
	/*
	 * If we booted from a floppy device, we ask about changing 
	 * floppies (so you can mount an alternate root disk)
	 * before continuing.
	 */
	if (B_TYPE(opendev) == FLOPPYDEV) {
		printf("File loaded -- change diskettes if necessary, then hit return");
		while (getchar() != '\n')
			;
	}
#endif

	i = (*((int (*)()) x.a_entry))(howto, opendev, cyloffset, bootparams);

	if (i)
		printf("exit %d\n", i); 
	return;
shread:
	printf("Short read\n");
	return;
}

void
makebootline()
{
	static char x[] = "0123456789abcdef";
	int param[4];
	char linebuf[sizeof line];
	register int i;
	register char *cp = linebuf;
	register char *cp2;

	param[0] = B_ADAPTOR(bootdev);
	param[1] = B_CONTROLLER(bootdev);
	param[2] = B_UNIT(bootdev);
	param[3] = B_PARTITION(bootdev);

	for (cp2 = devsw[B_TYPE(bootdev)].dv_name; *cp2; *cp++ = *cp2++)
		;

	*cp++ = '(';

	for (i = 0; i < sizeof param / sizeof param[0]; ++i) {
		if (i > 0)
			*cp++ = ',';
		if (param[i] >= 0xa) {
			*cp++ = '0';
			*cp++ = 'x';
		}
		if (param[i] >= 0x10) {
			*cp++ = x[param[i] >> 4];
			param[i] &= 0xf;
		}
		*cp++ = x[param[i]];
	}

	*cp++ = ')';

	for (cp2 = default_boot; *cp++ = *cp2++; )
		;

	for (cp = line, cp2 = linebuf; *cp++ = *cp2++; )
		;
}
